### PR TITLE
[ValueTracking] Compute KnownFP state from recursive select/phi.

### PR DESCRIPTION
**Pull request by davemgreen**
Given a recursive phi with select:
 %p = phi [ 0, entry ], [ %sel, loop]
 %sel = select %c, %other, %p

The fp state can be calculated using the knowledge that the select/phi pair can only be the initial state (0 here) or from %other. This adds a short-cut into computeKnownFPClass for PHI to detect that the select is recursive back to the phi, and if so use the state from the other operand.

This helps to address a regression from #83200.

### CODE REVIEW

**Review by pawosm-arm — APPROVED** [review]
Tested locally, this indeed restores the lost transformation.

**Comment by dtcxzyw** [comment]
I think this approach is also applicable to `computeKnownBits`. I'll post a patch later :)

**Comment by davemgreen** [comment]
> I think this approach is also applicable to `computeKnownBits`. I'll post a patch later :)

Nice, thanks! I had seen known bits had some handling of recurrences already, they didn't necessarily include selects.
I think we can expand this to phi+phi for the fpclass case too relatively easily, which will help in more complex cases.

I rebased over 577c7dd7cc4c5a9f62f9654cfa30ee9d55709426, which added a phase-ordering test.

**Comment by arsenm** [comment]
As far as the regression on #83200 is concerned, I think it needs a local flag handling fix

**Review by arsenm — COMMENTED** [review]
Needs a more targeted, dedicated test

**Comment by davemgreen** [comment]
> As far as the regression on https://github.com/llvm/llvm-project/pull/83200 is concerned, I think it needs a local flag handling fix

Yeah there are probably a few ways to fix it, selects are often missing fmfs due to them coming from control flow.  This seemed generally useful.

> Needs a more targeted, dedicated test

I've tried to add some Attributor tests, which I see have been used to test similar patches in the past. Let me know if there is a better way.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -6003,6 +6003,13 @@ void computeKnownFPClass(const Value *V, const APInt &DemandedElts,
         if (IncValue == P)
           continue;
 
+        // If the Use is a select of this phi, use the fp class of the other
+        // operand to break the recursion.
+        Value *V;
+        if (match(IncValue, m_Select(m_Value(), m_Specific(P), m_Value(V))) ||
+            match(IncValue, m_Select(m_Value(), m_Value(V), m_Specific(P))))
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:6010]: Just a nit, but maybe this means we should add `m_c_Select` matcher.

Comment (arsenm) [reply llvm/lib/Analysis/ValueTracking.cpp:6010]: It would be hard to represent that with the negated condition I think

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 1 above)
Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:6010]: Agreed, think it's probably not practical

**Pull request merged**